---
title: "1_preprocessing"
author: "Lola Giner Perez"
date: "8/6/2024"
output: html_document
---
```{r libraries}
library(phyloseq)
library(decontam)
library(tidyverse)
library(DT)
library(ggplot2); packageVersion("ggplot2")
```

#Identify Contaminants - Frequency
```{r Phyloseq}
ps_raw_ngless <- readRDS("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps_raw_ngless.Rds") 
df <- as.data.frame(sample_data(ps_raw_ngless)) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(ps_raw_ngless)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
ggplot(data=df, aes(x=Index, y=LibrarySize, color=Group)) + geom_point()
ggsave("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/Figures/contaminants_freq1.pdf")

contamdf.freq <- isContaminant(ps_raw_ngless, method="frequency", conc="Cantidad.de.DNA.total..ng.uL.")
head(contamdf.freq)
saveRDS(contamdf.freq, "/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/contaminant_table.Rds")
table(contamdf.freq$contaminant)
which(contamdf.freq$contaminant)
```
36 345 350 376 377 419 664 747 820 828 916 972

We have no negative control because it did not have enough DNA for sequencing, we cannot use Prevalence method.

The dashed black line of the plots shows the model of a noncontaminant sequence feature for which frequency is expected to be independent of the input DNA concentration. The red line shows the model of a contaminant sequence feature, for which frequency is expected to be inversely proportional to input DNA concentration, as contaminating DNA will make up a larger fraction of the total DNA in samples with very little total DNA. 
```{r plots}
set.seed(321)
plot_frequency(ps_raw_ngless, taxa_names(ps_raw_ngless)[c(36, 345, 350, 376, 377, 419, 664, 747, 820, 828, 916, 972)], conc="Cantidad.de.DNA.total..ng.uL.") + 
    xlab("DNA Concentration measured by Qubit")
ggsave("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/Figures/contaminants_freq2.pdf")
View(tax_table(ps_raw_ngless)[c(36, 345, 350, 376, 377, 419, 664, 747, 820, 828, 916, 972)])
```

Now that we have identified likely contaminants, letâ€™s remove them from the phyloseq object:
#```{r phyloseq 2}
#ps.noncontamfreq <- prune_taxa(!contamdf.freq$contaminant, ps)
#ps.noncontamfreq
#saveRDS(ps.noncontamfreq, "/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps.noncontamfreq.Rds")  
#```

Using the information that the decontam package gave me, I took the tax tables of what had been identified as contaminants and I made a deep search on the habitats of these taxa. Finally, I selected wich ones were correctly identifies as contaminants and now we will create a phyloseq object without the contaminant taxa.  
```{r phyloseq ok}
contaminants<-c("ref_mOTU_v31_03804", "ext_mOTU_v31_27169", "ref_mOTU_v31_05265", "ref_mOTU_v31_03970")
goodTaxa <- setdiff(taxa_names(ps_raw_ngless), contaminants)
ps.noncontam <- prune_taxa(goodTaxa, ps_raw_ngless)
ps.noncontam
saveRDS(ps.noncontam, "/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps.noncontam.Rds")  
```

#Data exploration
```{r Rarefaction curve}
library(readr)
library(data.table)

ps.noncontam <- readRDS("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps.noncontam.Rds")

#Rarefaction curve
library(vegan)
S <- specnumber(as.data.frame(t(otu_table(ps.noncontam)))) # observed number of species
(raremax <- min(rowSums(as.data.frame(t(otu_table(ps.noncontam))))))
Srare <- rarefy(t(as.data.frame(otu_table(ps.noncontam))), raremax)
plot(S, Srare, xlab = "Observed No. of Species", ylab = "Rarefied No. of Species")
abline(0, 1)
pdf("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/Figures/raref_curve.pdf")
rarecurve(as.data.frame(t(otu_table(ps.noncontam))), step = 20, sample = raremax, col = "black", cex = 0.6)
dev.off()
```

```{r exploration of the data}
sdt = data.table(as(sample_data(ps.noncontam), "data.frame"),
                 TotalReads = sample_sums(ps.noncontam), keep.rownames = TRUE)
pSeqDepth = ggplot(sdt, aes(TotalReads)) + geom_histogram() + ggtitle("Sequencing Depth")
pSeqDepth
pSeqDepth + facet_wrap(~Tipo)
ggsave("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/Figures/explore_data1.pdf")
pSeqDepth + facet_wrap(~Group)
ggsave("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/Figures/explore_data2.pdf")
```

###Exploration

```{r more exploration}
sdt = data.table(as(sample_data(ps.noncontam), "data.frame"),
                 TotalReads = sample_sums(ps.noncontam), keep.rownames = TRUE)
ggplot(sdt, aes(Cantidad.de.DNA.total..ng.uL., TotalReads, color = as.factor(Group))) + 
    geom_point(size = 5) + 
    geom_smooth(method = lm) +
    ggtitle("Sequencing Depth vs. DNAquant") +
    scale_y_log10() +
    facet_grid(Group ~ .)
ggsave("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/Figures/explore_data3.pdf")

tdt = data.table(tax_table(ps.noncontam),
                 TotalCounts = taxa_sums(ps.noncontam),
                 OTU = taxa_names(ps.noncontam))
ggplot(tdt, aes(TotalCounts)) + 
    geom_histogram() + 
    ggtitle("Histogram of Total Counts")
ggsave("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/Figures/explore_data4.pdf")
# How many singletons?
tdt[(TotalCounts <= 0), .N]
tdt[(TotalCounts <= 1), .N]
# None. This data has been filtered already. 
# How many doubletons?
tdt[(TotalCounts <= 2), .N]

#Filtering Threshold plot
# taxa cumulative sum
taxcumsum = tdt[, .N, by = TotalCounts]
setkey(taxcumsum, TotalCounts)
taxcumsum[, CumSum := cumsum(N)]
# Define the plot
pCumSum = ggplot(taxcumsum, aes(TotalCounts, CumSum)) + 
    geom_point() +
    xlab("Filtering Threshold, Minimum Total Counts") +
    ylab("OTUs Filtered") +
    ggtitle("OTUs that would be filtered vs. the minimum count threshold")
pCumSum
#Zoom in
pCumSum + xlim(0, 100)
ggsave("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/Figures/explore_data5.pdf")
```

#Rarefaction1
```{r rarefaction}
# keep result reproductive
 set.seed(321)
 ps.rarefied = rarefy_even_depth(ps.noncontam, rngseed=1, replace=F)
 ps.rarefied
 saveRDS(ps.rarefied, "/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps.rarefied.Rds")
```


#Taxa filtering: Remove low prevalent taxa
###For AphaDiv (keeping singletons)
It is important to not use filtered data because many richness estimates are modeled on singletons and doubletons in the occurrence table. So, you need to leave them in the dataset if you want a meaningful estimate.
Moreover, we usually not using normalized data because we want to assess the diversity on the raw data and we are not comparing samples to each other but only assessing diversity within each sample.
```{r ps_0}
# Remove samples with less than MINREADS from phyloseq object
ps_0 <- prune_taxa(taxa_sums(ps.noncontam) > 0, ps.noncontam)
# merge all taxa that are detected rare
total_samples <- phyloseq::nsamples(ps_0)
ps_0
saveRDS(ps_0, "/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps_0.Rds")
```

###For all the other tests
```{r tree plot}
library(microbiome)

##Create a prevalence dataframe 
Prevdf<- apply(X = otu_table(ps.noncontam),
               MARGIN = 1,
               FUN = function(x){sum(x > 0)})

##Add taxonomy and total read counts to this data.frame
Prevdf<- data.frame(Prevalence = Prevdf,
                    TotalAbundance = taxa_sums(ps.noncontam),
                    tax_table(ps.noncontam))

plyr::ddply(Prevdf, "Phylum", function(df1){
  data.frame(mean_prevalence=mean(df1$Prevalence),total_abundance=sum(df1$TotalAbundance,na.rm = T),stringsAsFactors = F)
})

ggplot(Prevdf, aes(TotalAbundance, Prevalence / nsamples(ps.noncontam),color=Phylum)) +
  geom_hline(yintercept = 0.1, alpha = 2, linetype = 2) + geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Log 10 Total Reads") + ylab("Prevalence [Prop. of Samples]") +
  theme_bw()+
  facet_wrap(~Phylum) + theme(legend.position="none")
ggsave("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/Figures/filter_data1.pdf")

#Remove low prevalent mOTUs (This data is used for differential abundance analysis)
metadata<- data.frame(sample_data(ps.noncontam)) 
data<-data.frame(sample_data(ps.noncontam))
samples<-data$SampleID
samples <- gsub("_[0-9]", "", samples)
data$samples<-samples
sample_data(ps.noncontam)<-data

sample_data <- sample_data(ps.noncontam)
sample_data$samples <- as.factor(sample_data$samples)

##Remove ASVs that do not show more than 5 times 
wh0 <- genefilter_sample(ps.noncontam, filterfun_sample(function(x) x > 5), A=0.01*nsamples(ps.noncontam))

# Create a logical vector for motus present in at least 3 volunteers (30% 3 out of 10)
asv_volunteer_filter <- apply(otu_table(ps.noncontam), 1, function(asv_counts) {
  # Find which volunteers have non-zero counts for this motu
  volunteers_with_asv <- unique(sample_data$samples[asv_counts > 0])
  # Check if motu is present in at least 3 volunteers
  length(volunteers_with_asv) >= 3
})

# Combine both filters
combined_filter <- wh0 & asv_volunteer_filter #Removing low prev taxa by % and if not present in at least 3/10 (30%) volunteers

ps_filtered <- prune_taxa(combined_filter, ps.noncontam)


pdf("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/Figures/filter_data2.pdf")
hist(readcount(ps_filtered))
dev.off()
saveRDS(ps_filtered, "/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps_filtered.Rds")
```

#Rarefaction2 after filtering
```{r rarefaction}
# keep result reproductive
 set.seed(321)
 ps.rarefied.filtered = rarefy_even_depth(ps_filtered, rngseed=1, replace=F)
 ps.rarefied.filtered
 saveRDS(ps.rarefied.filtered, "/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps_rarefied_filtered.Rds")
```

#Data transformation for Beta Div
```{r rarefaction}
ps_comp<- microbiome::transform(ps_filtered, "compositional")
saveRDS(ps_comp, "/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps_comp.Rds")

ps_clr<- microbiome::transform(ps_filtered, "clr")
saveRDS(ps_clr, "/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps_clr.Rds")

ps_comp_raref<- microbiome::transform(ps.rarefied.filtered, "compositional")
saveRDS(ps_comp_raref, "/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps_comp_raref.Rds")

ps_clr_raref<- microbiome::transform(ps.rarefied.filtered, "clr")
saveRDS(ps_clr_raref, "/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/ps_clr_raref.Rds")
```


```{r Rarefaction curve}
library(readr)
library(data.table)

#Rarefaction curve
library(vegan)
S <- specnumber(as.data.frame(t(otu_table(ps_filtered)))) # observed number of species
(raremax <- min(rowSums(as.data.frame(t(otu_table(ps_filtered))))))
Srare <- rarefy(t(as.data.frame(otu_table(ps_filtered))), raremax)
plot(S, Srare, xlab = "Observed No. of Species", ylab = "Rarefied No. of Species")
abline(0, 1)
pdf("/fast/AG_Forslund/Lola/Secuencias_INCLIVA_2024/MHE_rif/outputs/ngless/Figures/raref_curve_filtered.pdf")
rarecurve(as.data.frame(t(otu_table(ps_filtered))), step = 20, sample = raremax, col = "black", cex = 0.6)
dev.off()
```